src/core/grid_based_algorithms/lattice.cpp

```diff
@@ -117,7 +116,7 @@ void Lattice::map_position_to_lattice(const Utils::Vector3d &pos,
 
 Utils::Vector3i Lattice::local_index(Utils::Vector3i const &global_index) const
     noexcept {
-  return {global_index[0] - node_pos[0] * grid[0] + halo_size,
-          global_index[1] - node_pos[1] * grid[1] + halo_size,
-          global_index[2] - node_pos[2] * grid[2] + halo_size};
+  return {global_index[0] - local_index_offset[0] + halo_size,
```

Comment:
we are almost there... having a readable ```Lattice``` class

Reply:
Maybe it will go away soon anyway: the p3m contains similar logic, but with more features. If there is a quick way to reuse that I'll do it. That also includes halo comm, so the lb halo stuff (for the coupling) would also go away. Let's see.